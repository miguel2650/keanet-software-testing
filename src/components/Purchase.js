import React from "react";
import CellPhonesTable from "./CellPhonesTable";

class Purchase extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      internetConnection: false,
      phoneLines: 0,
      cellPhones: [],
      price: 0
    };
    //this.includeInternet = this.includeInternet.bind(this);
  }

  priceList = {
    "Motorola G99": 800,
    "iPhone 99": 6000,
    "Samsung Galaxy 99": 1000,
    "Sony Xperia 99": 900,
    "Huawei 99": 900
  };

  includeInternet = input => {
    if (typeof input !== "boolean") {
      return () => {
        throw new TypeError("Wrong type");
      };
    } else {
      this.setState({ internetConnection: input });
      if (input) {
        this.setState({ price: this.state.price + 200 });
        return this.state.price;
      } else {
        this.setState({ price: this.state.price - 200 });
        return this.state.price;
      }
    }
  };

  addPhoneLine = () => {
    if (this.state.phoneLines >= 8) {
      return () => {
        throw new Error("Can not exceed value of 8");
      };
    } else {
      this.setState({
        phoneLines: this.state.phoneLines + 1
      });
      this.setState({
        price: this.state.price + 150
      });
    }
  };

  removePhoneLine = () => {
    if (this.state.phoneLines <= 0) {
      return () => {
        throw new Error("Can not subceed value of 0");
      };
    } else {
      this.setState({ phoneLines: this.state.phoneLines - 1 });
      this.setState({ price: this.state.price - 150 });
    }
  };

  selectCellPhone = modelName => {
    if (typeof modelName !== "string") {
      return () => {
        throw new TypeError("Wrong type");
      };
    } else {
      if (modelName in this.priceList) {
        const newList = [...this.state.cellPhones];
        newList.push(modelName);
        this.setState({ cellPhones: newList });
        this.setState({ price: this.state.price + this.priceList[modelName] });
      }
    }
    return this.state.price;
  };

  deselectCellPhone = modelName => {
    if (this.state.cellPhones.includes(modelName)) {
      const newList = [...this.state.cellPhones];
      const index = newList.indexOf(modelName);
      newList.splice(index, 1);
      this.setState({ cellPhones: newList });
      this.setState({ price: this.state.price - this.priceList[modelName] });
    }
    return this.state.price;
  };

  buying = () => {
    if (this.state.price <= 0) {
      return "Please select some items before checking out";
    } else {
      let buyMessage = "";
      if (this.state.internetConnection) {
        buyMessage += "1x Internet Connection = 200 DKK (monthly) \n";
      }
      if (this.state.phoneLines > 0) {
        buyMessage += `${this.state.phoneLines}x Phone Lines = ${
          this.state.phoneLines * 150
        } DKK (monthly)\n`;
      }
      if (this.state.cellPhones.length > 0) {
        var counts = {};
        for (var i = 0; i < this.state.cellPhones.length; i++) {
          counts[this.state.cellPhones[i]] =
            1 + (counts[this.state.cellPhones[i]] || 0);
        }
        for (const phone in counts) {
          buyMessage += `${counts[phone]}x ${phone} = ${
            counts[phone] * this.priceList[phone]
          } DKK \n`;
        }
      }
      buyMessage += ` \n You bought for a total of: ${this.state.price} DKK \n`;
      return buyMessage;
    }
  };

  render() {
    return (
      <React.Fragment>
        <h1>KEANet</h1>
        <input
          id="includeInternet"
          type="checkbox"
          onClick={() => this.includeInternet(!this.state.internetConnection)}
        ></input>
        <label>Internet Connection</label>
        <br></br>
        <label>Phone Lines </label>
        <input
          id="phoneLines"
          type="number"
          max="8"
          min="0"
          onPaste={e => {
            window.alert("Please do not paste...");
            e.preventDefault();
          }}
          onKeyDown={e => e.preventDefault()}
          onAuxClick={() =>
            window.alert(
              "Please use the arrows instead of trying to break the system"
            )
          }
          onChange={e => {
            if (e.target.value > this.state.phoneLines) {
              this.addPhoneLine();
            } else {
              this.removePhoneLine();
            }
          }}
        ></input>
        <br></br>
        <CellPhonesTable
          selectCellPhone={this.selectCellPhone}
          deselectCellPhone={this.deselectCellPhone}
          cellPhones={this.state.cellPhones}
          priceList={this.priceList}
        ></CellPhonesTable>
        <label>Total price: {this.state.price} DKK</label>
        <br></br>
        <button
          onClick={() => {
            window.confirm(this.buying());
          }}
        >
          Buy
        </button>
      </React.Fragment>
    );
  }
}
export default Purchase;
