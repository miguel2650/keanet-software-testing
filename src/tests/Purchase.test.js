import React from "react";
import { render } from "@testing-library/react";
import Purchase from "../components/Purchase";
import { shallow } from "enzyme";

test("renders KEANet <h1> tag", () => {
  const { getByText } = render(<Purchase></Purchase>);
  const linkElement = getByText("KEANet");
  expect(linkElement).toBeInTheDocument();
});

// Using enzyme to simulate
const wrapper = shallow(<Purchase></Purchase>);

// This test is a Enzyme and Jest specific test. This creates the
// wrapper components as HTML and makes sure that is is rendered probably.
it("should match the snapshot", () => {
  expect(wrapper.html()).toMatchSnapshot();
});

// Testing the actual side effects of click the checkbox
describe("includeInternet function", () => {
  it("should change state of internetConnection to true when checkbox is clicked", () => {
    expect(wrapper.state("internetConnection")).toEqual(false);
    wrapper.find("#includeInternet").simulate("click");
    expect(wrapper.state("internetConnection")).toEqual(true);
    // If not simulated once again, the state will remain price: 200, internetConnection: true
    wrapper.find("#includeInternet").simulate("click");
  });
});

// Testing the parameters passed to includeInternet() function
describe("includeInternet()", () => {
  it("should fail at all datatypes but boolean", () => {
    //let component = renderer.create(<Purchase></Purchase>).getInstance();
    // Data for inputs with expected outputs
    let mockData = {
      case0: {
        input: true,
        expected: 200
      },
      case1: {
        input: false,
        expected: 0
      },
      case2: {
        input: "stringlala",
        expected: TypeError("Wrong type")
      },
      case3: {
        input: "--",
        expected: TypeError("Wrong type")
      },
      case4: {
        input: 0,
        expected: TypeError("Wrong type")
      },
      case5: {
        input: null,
        expected: TypeError("Wrong type")
      }
    };
    let counter = 0;
    for (let testCase in mockData) {
      console.log("Test case: ", counter);
      const instance = wrapper
        .instance()
        .includeInternet(mockData[testCase].input);
      // Test cases where error is expected
      if (counter >= 2) {
        expect(instance).toThrow(mockData[testCase].expected);
      } else {
        // Text cases where value is expected
        expect(instance).toBe(mockData[testCase].expected);
      }
      counter++;
    }
  });

  // Testing the actual side effects of click to addPhoneLine()
  describe("addPhoneLine() function", () => {
    it("should change state of phoneLines and price when phoneLine is added", () => {
      const instance = wrapper.instance();
      expect(wrapper.state("phoneLines")).toBe(0);
      instance.addPhoneLine();
      expect(wrapper.state("phoneLines")).toBe(1);
    });
  });

  describe("removePhoneLine() function", () => {
    it("should change state of phoneLines and price when phoneLine is removed", () => {
      const instance = wrapper.instance();
      expect(wrapper.state("phoneLines")).toBe(1);
      instance.removePhoneLine();
      expect(wrapper.state("phoneLines")).toBe(0);
      instance.removePhoneLine();
      expect(wrapper.state("phoneLines")).toBe(-1);
    });
  });
});

/*

TODO: We are unable to test for type errors
Look inside Purchase.JS @ includeInternet()
The input of whatever type will never crash the application.
Any type of input can be acceptable.
Changing to TypeScript can fix our issue or adding some kind of checking mechanism.
Like: If (typeOf(input) === Boolean) {} else {throw new Error()}
Give me your feedback.

------------------------------------------------------
Dont mind this, was used for testing a error testcase.

      let tree2 = component.includeInternet("string");
      expect(tree2).toThrow();

      case3: {
        input: "thisIsString",
        expected: new Error("Eror")
      }
*/
