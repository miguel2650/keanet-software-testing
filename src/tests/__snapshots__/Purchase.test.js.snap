// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `"<h1>KEANet</h1><input type=\\"checkbox\\" id=\\"includeInternet\\"/><label>Internet Connection</label><br/><label>Phone Lines </label><input type=\\"number\\" id=\\"phoneLines\\" max=\\"8\\" min=\\"0\\"/><br/><label>Cell phones</label><table><tbody><tr><td><select size=\\"5\\" id=\\"cellPhones\\" style=\\"width:200px\\"><option value=\\"Motorola G99\\">Motorola G99</option><option value=\\"iPhone 99\\">iPhone 99</option><option value=\\"Samsung Galaxy 99\\">Samsung Galaxy 99</option><option value=\\"Sony Xperia 99\\">Sony Xperia 99</option><option value=\\"Huawei 99\\">Huawei 99</option></select></td><td><button>&gt;</button><br/><button>&lt;</button></td><select size=\\"5\\" id=\\"selectedPhones\\" style=\\"width:200px\\"></select></tr></tbody></table><label>Total price: 0 DKK</label><br/><button>Buy</button>"`;
